File: app/src/main/java/sgtmelon/scriptum/element/SortDialog.java
package sgtmelon.scriptum.element;

import android.app.Dialog;
import android.os.Bundle;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SimpleItemAnimator;
import sgtmelon.safedialog.office.annot.DialogAnn;
import sgtmelon.safedialog.office.blank.DialogBlank;
import sgtmelon.scriptum.R;
import sgtmelon.scriptum.app.adapter.SortAdapter;
import sgtmelon.scriptum.app.model.item.SortItem;
import sgtmelon.scriptum.office.annot.def.SortDef;
import sgtmelon.scriptum.office.intf.ItemIntf;
import sgtmelon.scriptum.office.utils.PrefUtils;

public final class SortDialog extends DialogBlank implements ItemIntf.ClickListener {

    // TODO: 07.10.2018 От частного к общему

    private final List<SortItem> listSort = new ArrayList<>();

    private String init, keys;
    private String[] text;
    private SortAdapter adapter;

    private final ItemTouchHelper.Callback touchCallback = new ItemTouchHelper.Callback() {
        @Override
        public int getMovementFlags(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
            int flagsDrag = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
            int flagsSwipe = 0;

            return makeMovementFlags(flagsDrag, flagsSwipe);
        }

        @Override
        public void clearView(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
            super.clearView(recyclerView, viewHolder);

            adapter.notifyItemChanged(viewHolder.getAdapterPosition());
        }

        @Override
        public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {

        }

        @Override
        public boolean onMove(@NonNull RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
            final int positionFrom = viewHolder.getAdapterPosition();
            final int positionTo = target.getAdapterPosition();

            final SortItem sortItem = listSort.get(positionFrom);

            listSort.remove(positionFrom);
            listSort.add(positionTo, sortItem);

            adapter.setList(listSort);
            adapter.notifyItemMoved(positionFrom, positionTo);

            adapter.notifyItemChanged(positionFrom == adapter.sortSt.getEnd()
                    ? positionTo
                    : positionFrom);

            keys = PrefUtils.getSortByList(listSort);
            setEnable();
            return true;
        }
    };

    public void setArguments(String keys) {
        final Bundle bundle = new Bundle();

        bundle.putString(DialogAnn.INIT, keys);
        bundle.putString(DialogAnn.VALUE, keys);

        setArguments(bundle);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final Bundle bundle = getArguments();
        if (savedInstanceState != null) {
            init = savedInstanceState.getString(DialogAnn.INIT);
            keys = savedInstanceState.getString(DialogAnn.VALUE);
        } else if (bundle != null) {
            init = bundle.getString(DialogAnn.INIT);
            keys = bundle.getString(DialogAnn.VALUE);
        }

        text = getResources().getStringArray(R.array.pref_sort_text);

        final RecyclerView recyclerView = new RecyclerView(context);

        final int padding = context.getResources().getInteger(R.integer.dlg_recycler_padding);
        recyclerView.setPadding(padding, padding, padding, padding);
        recyclerView.setOverScrollMode(View.OVER_SCROLL_NEVER);

        recyclerView.setLayoutManager(new LinearLayoutManager(context));

        adapter = new SortAdapter(context);
        adapter.setClickListener(this);

        listSort.clear();
        for (String aKey : keys.split(SortDef.divider)) {
            final int key = Integer.parseInt(aKey);
            final SortItem sortItem = new SortItem(text[key], key);
            listSort.add(sortItem);
        }
        adapter.setList(listSort);

        recyclerView.setAdapter(adapter);

        final ItemTouchHelper itemTouchHelper = new ItemTouchHelper(touchCallback);
        itemTouchHelper.attachToRecyclerView(recyclerView);

        final SimpleItemAnimator animator = (SimpleItemAnimator) recyclerView.getItemAnimator();
        if (animator != null) {
            animator.setSupportsChangeAnimations(false);
        }

        return new AlertDialog.Builder(context)
                .setTitle(getString(R.string.dialog_title_sort))
                .setView(recyclerView)
                .setPositiveButton(getString(R.string.dialog_btn_accept), onPositiveClick)
                .setNegativeButton(getString(R.string.dialog_btn_cancel), (dialog, id) -> dialog.cancel())
                .setNeutralButton(getString(R.string.dialog_btn_reset), onNeutralClick)
                .setCancelable(true)
                .create();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putString(DialogAnn.INIT, init);
        outState.putString(DialogAnn.VALUE, keys);
    }

    public String getKeys() {
        return keys;
    }

    @Override
    protected void setEnable() {
        super.setEnable();

        buttonPositive.setEnabled(!PrefUtils.getSortEqual(init, keys));
        buttonNeutral.setEnabled(!PrefUtils.getSortEqual(SortDef.def, keys));
    }

    @Override
    public void onItemClick(View view, int p) {
        final SortItem sortItem = listSort.get(p);

        final int key = sortItem.getKey() == SortDef.create
                ? SortDef.change
                : SortDef.create;

        sortItem.setText(text[key]);
        sortItem.setKey(key);

        adapter.setListItem(p, sortItem);
        adapter.notifyItemChanged(p);

        keys = PrefUtils.getSortByList(listSort);
        setEnable();
    }

}

File: app/src/main/java/sgtmelon/scriptum/element/SortDialog.kt
package sgtmelon.scriptum.element

import android.app.Dialog
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SimpleItemAnimator
import sgtmelon.safedialog.office.annot.DialogAnn
import sgtmelon.safedialog.office.blank.DialogBlank
import sgtmelon.scriptum.R
import sgtmelon.scriptum.app.adapter.SortAdapter
import sgtmelon.scriptum.app.model.item.SortItem
import sgtmelon.scriptum.office.annot.def.SortDef
import sgtmelon.scriptum.office.intf.ItemIntf
import sgtmelon.scriptum.office.utils.HelpUtils
import sgtmelon.scriptum.office.utils.HelpUtils.Sort.getSort

class SortDialog : DialogBlank(), ItemIntf.ClickListener {

    private val listSort: MutableList<SortItem> = ArrayList()

    private var init: String = ""

    var keys: String = ""
        private set

    private lateinit var text: Array<String>
    private lateinit var adapter: SortAdapter

    private val touchCallback = object : ItemTouchHelper.Callback() {
        override fun getMovementFlags(recyclerView: RecyclerView,
                                      viewHolder: RecyclerView.ViewHolder): Int {
            val flagsDrag = ItemTouchHelper.UP or ItemTouchHelper.DOWN
            val flagsSwipe = 0

            return ItemTouchHelper.Callback.makeMovementFlags(flagsDrag, flagsSwipe)
        }

        override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
            super.clearView(recyclerView, viewHolder)

            adapter.notifyItemChanged(viewHolder.adapterPosition)
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {

        }

        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder,
                            target: RecyclerView.ViewHolder): Boolean {
            val positionFrom = viewHolder.adapterPosition
            val positionTo = target.adapterPosition

            val sortItem = listSort[positionFrom]

            listSort.removeAt(positionFrom)
            listSort.add(positionTo, sortItem)

            adapter.setList(listSort)
            adapter.notifyItemMoved(positionFrom, positionTo)

            val position = if (positionFrom == adapter.sortSt.end) positionTo
            else positionFrom

            adapter.notifyItemChanged(position)

            keys = HelpUtils.Sort.getSortByList(listSort)

            setEnable()
            return true
        }
    }

    fun setArguments(keys: String) {
        val bundle = Bundle()

        bundle.putString(DialogAnn.INIT, keys)
        bundle.putString(DialogAnn.VALUE, keys)

        arguments = bundle
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        init = savedInstanceState?.getString(DialogAnn.INIT)
                ?: arguments?.getString(DialogAnn.INIT)
                ?: ""

        keys = savedInstanceState?.getString(DialogAnn.VALUE)
                ?: arguments?.getString(DialogAnn.VALUE)
                ?: ""

        text = resources.getStringArray(R.array.pref_sort_text)

        val recyclerView = RecyclerView(activity)

        val padding = activity.resources.getInteger(R.integer.dlg_recycler_padding)
        recyclerView.setPadding(padding, padding, padding, padding)
        recyclerView.overScrollMode = View.OVER_SCROLL_NEVER

        recyclerView.layoutManager = LinearLayoutManager(context)

        adapter = SortAdapter(activity, this)

        listSort.clear()
        for (aKey in keys.split(SortDef.divider.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()) {
            val key = Integer.parseInt(aKey)
            val sortItem = SortItem(text[key], key)
            listSort.add(sortItem)
        }

        adapter.setList(listSort)

        recyclerView.adapter = adapter

        val itemTouchHelper = ItemTouchHelper(touchCallback)
        itemTouchHelper.attachToRecyclerView(recyclerView)

        val animator = recyclerView.itemAnimator as SimpleItemAnimator?
        if (animator != null) {
            animator.supportsChangeAnimations = false
        }

        return AlertDialog.Builder(activity)
                .setTitle(getString(R.string.dialog_title_sort))
                .setView(recyclerView)
                .setPositiveButton(getString(R.string.dialog_btn_accept), onPositiveClick)
                .setNegativeButton(getString(R.string.dialog_btn_cancel)) { dialog, _ -> dialog.cancel() }
                .setNeutralButton(getString(R.string.dialog_btn_reset), onNeutralClick)
                .setCancelable(true)
                .create()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putString(DialogAnn.INIT, init)
        outState.putString(DialogAnn.VALUE, keys)
    }

    override fun setEnable() {
        super.setEnable()

        buttonPositive.isEnabled = !HelpUtils.Sort.getSortEqual(init, keys)
        buttonNeutral.isEnabled = !HelpUtils.Sort.getSortEqual(SortDef.def, keys)
    }

    override fun onItemClick(view: View, p: Int) {
        if (p == RecyclerView.NO_POSITION || p != adapter.sortSt.end) return

        val sortItem = listSort[p]

        val key = if (sortItem.key == SortDef.create) SortDef.change
        else SortDef.create

        sortItem.text = text[key]
        sortItem.key = key

        adapter.setListItem(p, sortItem)
        adapter.notifyItemChanged(p)

        keys = listSort.getSort()
        setEnable()
    }

}


